Add support for creating reproducers for the LTP skipfile

katieworton opened the pull request:
Add support for rerunning the tests in the LTP skipfile via the `squad-create-skipfile-reproducers` script.

By default, if only the `group` parameter is provided, the script will create a TuxPlan of reproducers for all supported QEMU devices and branches.

Builds on https://github.com/Linaro/squad-client-utils/pull/21

katieworton committed:
Add documentation for skipfile reproducers

Add usage of squad-create-skipfile-reproducers to the README.md

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton commented:
Code is ready for review @roxell

roxell commented:
When I run this command I see the following traceback
./squad-create-skipfile-reproducers --group lkft --branch linux-next --device-name fvp-aemva --build-names gcc-12-lkftconfig --suite-name ltp-syscalls
```
Traceback (most recent call last):
  File "/home/anders/src/squad-client-utils/./squad-create-skipfile-reproducers", line 251, in <module>
    sys.exit(run())
             ^^^^^
  File "/home/anders/src/squad-client-utils/./squad-create-skipfile-reproducers", line 240, in run
    reproducer = squadutilslib.create_tuxsuite_plan_from_tuxsuite_tests(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/anders/src/squad-client-utils/squadutilslib.py", line 282, in create_tuxsuite_plan_from_tuxsuite_tests
    plan = yaml.load(test_yaml_str)
           ^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: load() missing 1 required positional argument: 'Loader'
```

roxell commented on squad-create-skipfile-reproducers:
```diff
@@ -0,0 +1,251 @@
+                    if branch_name in args.branches:
+                        projects.append(look_up_project_by_branch(branch_name))
+                    else:
+                        logger.warning(
```
Can probably be a logger.debug

katieworton replied:
Done

roxell commented on squad-create-skipfile-reproducers:
```diff
@@ -0,0 +1,251 @@
+    "linux-5.10.y": "linux-stable-rc",
+    "linux-5.15.y": "linux-stable-rc",
+    "linux-6.1.y": "linux-stable-rc",
+    "linux-6.2.y": "linux-stable-rc",
```
linux-6.3.y?

katieworton replied:
Changed as a temp fix before adding auto generation of this list.

roxell commented on squad-create-skipfile-reproducers:
```diff
@@ -0,0 +1,251 @@
+linux_mainline_next_branches = {
+    "linux-mainline": "master",
+    "linux-next": "master",
+}
```
As we spoke about we should probably go the other way around, look in SQUAD after projects that have been updated in LKFT's group the last 3 weeks. pull them and don't hardcode branches/projects like this.

katieworton replied:
Noted - will experiment with this and then perform the update as discussed. Should be much easier to maintain the code this way.

katieworton commented:
> When I run this command I see the following traceback ./squad-create-skipfile-reproducers --group lkft --branch linux-next --device-name fvp-aemva --build-names gcc-12-lkftconfig --suite-name ltp-syscalls
> 
> ```
> Traceback (most recent call last):
>   File "/home/anders/src/squad-client-utils/./squad-create-skipfile-reproducers", line 251, in <module>
>     sys.exit(run())
>              ^^^^^
>   File "/home/anders/src/squad-client-utils/./squad-create-skipfile-reproducers", line 240, in run
>     reproducer = squadutilslib.create_tuxsuite_plan_from_tuxsuite_tests(
>                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/anders/src/squad-client-utils/squadutilslib.py", line 282, in create_tuxsuite_plan_from_tuxsuite_tests
>     plan = yaml.load(test_yaml_str)
>            ^^^^^^^^^^^^^^^^^^^^^^^^
> TypeError: load() missing 1 required positional argument: 'Loader'
> ```

"Works on my PC :shrug: " ;) 

Things seem to work in my environment but looks like we are probably using different pyyaml versions. Easy to fix (by picking and setting a pyyaml version requirement; adding the loader parameter to `load()` or using `full_load()` rather than `load()`).

I added the parameter `Loader=yaml.FullLoader` as the current fix.

As an extra point here - we might want to use the ruamel.yaml library instead. This library has been picked for the later patches that deal with updating the skipfiles as it has features that we need but pyyaml doesn't support. Perhaps it doesn't make sense to be using two different yaml parsers in the same project.

katieworton commented:
Other changes made outside of review comments:
- Added skipfile URL as parameter
- Changed “temp” to “tmp” in variable names for consistency with project, and other minor variable renaming
- Return -1 if a reproducer is not found rather than skipping it
- Removed unnecessary “tests” entry in yaml file
- Added the group name to the plan file name
- Changed to print errors to the terminal rather than to a file.

roxell commented:
overlays should be a list

roxell commented:
$ tuxsuite plan skipfile-reproducer-lkft-linux-next-master-plan.yaml
Running Linux Kernel plan 'skipfile-reproducer-lkft-linux-next-master-plan.yaml': 'Run tests from customised reproducers.'
Plan https://tuxapi.tuxsuite.com/v1/groups/linaro/projects/anders/plans/2Rn2PMNxGVvaq5gG1VSR0Xxz4bP

uid: 2Rn2PMNxGVvaq5gG1VSR0Xxz4bP
⚙️  Provisioning: [boot] qemu-i386 @ https://tuxapi.tuxsuite.com/v1/groups/linaro/projects/anders/tests/2Rn2PMNU1EDdrw4O4Ph2bKS25r3
⚙️  Provisioning: [boot] qemu-i386 @ https://tuxapi.tuxsuite.com/v1/groups/linaro/projects/anders/tests/2Rn2PLhtuVdurJx4wBTSYGSdedL
🔧 TuxTest error: [boot] qemu-i386 @ https://tuxapi.tuxsuite.com/v1/groups/linaro/projects/anders/tests/2Rn2PMNU1EDdrw4O4Ph2bKS25r3
🔧 TuxTest error: [boot] qemu-i386 @ https://tuxapi.tuxsuite.com/v1/groups/linaro/projects/anders/tests/2Rn2PLhtuVdurJx4wBTSYGSdedL

Summary: https://tuxapi.tuxsuite.com/v1/groups/linaro/projects/anders/plans/2Rn2PMNxGVvaq5gG1VSR0Xxz4bP

katieworton commented:
> $ tuxsuite plan skipfile-reproducer-lkft-linux-next-master-plan.yaml Running Linux Kernel plan 'skipfile-reproducer-lkft-linux-next-master-plan.yaml': 'Run tests from customised reproducers.' Plan https://tuxapi.tuxsuite.com/v1/groups/linaro/projects/anders/plans/2Rn2PMNxGVvaq5gG1VSR0Xxz4bP
> 
> uid: 2Rn2PMNxGVvaq5gG1VSR0Xxz4bP ⚙️ Provisioning: [boot] qemu-i386 @ https://tuxapi.tuxsuite.com/v1/groups/linaro/projects/anders/tests/2Rn2PMNU1EDdrw4O4Ph2bKS25r3 ⚙️ Provisioning: [boot] qemu-i386 @ https://tuxapi.tuxsuite.com/v1/groups/linaro/projects/anders/tests/2Rn2PLhtuVdurJx4wBTSYGSdedL 🔧 TuxTest error: [boot] qemu-i386 @ https://tuxapi.tuxsuite.com/v1/groups/linaro/projects/anders/tests/2Rn2PMNU1EDdrw4O4Ph2bKS25r3 wrench TuxTest error: [boot] qemu-i386 @ https://tuxapi.tuxsuite.com/v1/groups/linaro/projects/anders/tests/2Rn2PLhtuVdurJx4wBTSYGSdedL
> 
> Summary: https://tuxapi.tuxsuite.com/v1/groups/linaro/projects/anders/plans/2Rn2PMNxGVvaq5gG1VSR0Xxz4bP

It seems like the change in the quotes required by tuxsuite broke this PR's current's code use of '"..."' around the commands. This error appears to be fixed by updating the quotes to '...'. I will update the tuxsuite code to use single quotes (the same as TuxRun uses).

However, there also seems to be an issue where the `boot-args` parameter is being dropped and doesn't appear in the tuxtest and tuxrun reproducers https://tuxapi.tuxsuite.com/v1/groups/linaro/projects/katie/tests/2Ro7n69rDIon9u4UZIhVp6uTCZw
I have asked about this in comments of the ticket adding support for single quotes https://linaro.atlassian.net/browse/STG-4704

katieworton committed:
Add support for creating LTP skipfile reproducers

Add the squad-create-skipfile-reproducers script to create TuxRun
reproducers for testing the LTP skipfile entries.
By providing the group and optionally the branches, devices and build
names, the script will create reproducers for all combinations of these
parameters. If any of these parameters are not provided, the default
behaviour is to run will all supported values for the given parameter.
Extra parameters include:
  * a --allow-unfinished flag to allow fetching of
    reproducers where the build has not been marked as finished.
  * A --suite-name parameter to set which LTP suite reproducer should be
    fetched when creating reproducers.
  * A --debug flag to add extra debug prints when the script runs.

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton committed:
Add support for TuxPlan reproducers

Add support for creating TuxPlan reproducers from a list of TuxTest
reproducers. Add this as the default option of skipfile reproducers. Add
a --local flag to create TuxRun reproducers.

Signed-off-by: Katie Worton <katie.worton@linaro.org>

roxell commented on squad-create-skipfile-reproducers:
```diff
@@ -0,0 +1,237 @@
+        if pathlib.Path(tmp_reproducer_filename).exists():
+            pathlib.Path.unlink(pathlib.Path(tmp_reproducer_filename))
+
+    logger.info("Finished creating skipfile reproducers.")
```
should probably print the filename here so we know.

katieworton replied:
Done

roxell commented on squad-create-skipfile-reproducers:
```diff
@@ -0,0 +1,237 @@
+            pathlib.Path.unlink(pathlib.Path(tmp_reproducer_filename))
+
+    logger.info("Finished creating skipfile reproducers.")
+    logger.info(f"Took {time.time() - start}s")
```
should probably be debug and not info.

katieworton replied:
Done

roxell commented on squad-create-skipfile-reproducers:
```diff
@@ -0,0 +1,237 @@
+                    if branch_name in args.branches:
+                        projects.append(look_up_project_by_branch(branch_name))
+                    else:
+                        logger.debug(
```
Can't this be done for every branch and we can drop the "else:" bit?

katieworton replied:
Not sure I understand what you mean here - are you talking about the debug print that prints when a branch from the skipfile entry is skipped?

roxell commented on squad-create-skipfile-reproducers:
```diff
@@ -0,0 +1,237 @@
+    with open(skipfile) as f:
+        reasons = yaml.load(f, Loader=yaml.FullLoader)
+    for reason in reasons["skiplist"]:
+        if reason["boards"] == "all" or "all" in reason["boards"]:
```
if no boards or a specific board what do we do then?

katieworton replied:
Added check for cases where the list in the skipfile entry and our boards list have common members

roxell commented on squadutilslib.py:
```diff
@@ -185,6 +186,7 @@ def get_reproducer(
         except HTTPError:
             logger.error(f"Reproducer not found at {testrun.job_url}!")
             raise ReproducerNotFound
+
```
probably not needed in this patch

katieworton replied:
Should be fixed now :+1:

roxell commented on squadutilslib.py:
```diff
@@ -229,3 +231,63 @@ def create_custom_reproducer(
+
+
+def create_tuxsuite_plan_from_tuxsuite_tests(tuxtest_filename, plan_name):
+    tuxrun_list = open(tuxtest_filename).read().splitlines()
```
tuxrun_list is missleading since you grab a tuxtest reproducer right?

katieworton replied:
Renamed to tuxtest_list

katieworton committed:
Fix variable naming for tuxrun/tuxtest reproducers

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton committed:
Change import style to limit imports squadutilslib, squad-create-reproducer

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton committed:
Change reproducer functions to return string rather than file name

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton committed:
Change squadutilslib logging to debug prints

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton commented:
Updates:

- Renamed "tuxrun" variable to "reproducer" as it is used for both tuxrun and tuxtest reproducers
- Changed import style from `import X` to `from X import Y` to minimise imports
- Changed reproducer functions to return strings rather than the file name - allowed for less temporary files to be created when generating the skipfile reproducers
- Changed some `info` logging to `debug` to reduce noise in outputs

katieworton commented:
@roxell see updates - should be ready for review :+1:

roxell commented on squad-create-skipfile-reproducers:
```diff
@@ -0,0 +1,264 @@
+    "linux-5.10.y": "linux-stable-rc",
+    "linux-5.15.y": "linux-stable-rc",
+    "linux-6.1.y": "linux-stable-rc",
+    "linux-6.3.y": "linux-stable-rc",
```
this should be linux-6.4.y

roxell commented:
This is what I see
$ ./squad-create-skipfile-reproducers --group lkft --branch linux-next --device-name qemu-i386 --build-names gcc-12-lkftconfig --suite-name ltp-syscalls
ERROR:__main__:No reproducer could be found for lkft linux-next-master qemu-i386 ['gcc-12-lkftconfig'] 

But I can see a job from tag next-20230714
https://qa-reports.linaro.org/lkft/linux-next-master/build/next-20230714/testrun/18305409/suite/ltp-syscalls/test/sendfile07/details/

roxell commented:
> This is what I see $ ./squad-create-skipfile-reproducers --group lkft --branch linux-next --device-name qemu-i386 --build-names gcc-12-lkftconfig --suite-name ltp-syscalls ERROR:**main**:No reproducer could be found for lkft linux-next-master qemu-i386 ['gcc-12-lkftconfig']
> 
> But I can see a job from tag next-20230714 https://qa-reports.linaro.org/lkft/linux-next-master/build/next-20230714/testrun/18305409/suite/ltp-syscalls/test/sendfile07/details/

'--allow-unfinished' fixes the issue =)
